File: java/solitaire/klondike/src/main/java/org/silnith/game/solitaire/move/FoundationToColumnMove.java
package org.silnith.game.solitaire.move;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.ListIterator;
import java.util.Map;

import org.silnith.deck.Card;
import org.silnith.deck.Suit;
import org.silnith.game.solitaire.Board;
import org.silnith.game.solitaire.Column;
import org.silnith.util.Pair;


/**
 * A move that takes a card from the Foundation and puts it onto a column run.
 * 
 * <p>If the column is empty, only a {@link org.silnith.deck.Value#KING} may be
 * moved.  If the column is not empty, then the card must adhere to the rules of
 * a run.</p>
 */
public class FoundationToColumnMove implements SolitaireMove {
	
	/**
	 * Finds all moves from the foundation to a column run.
	 * 
	 * @param board the board to examine
	 * @return a collection of moves
	 */
	public static Collection<FoundationToColumnMove> findMoves(final Board board) {
		final Collection<FoundationToColumnMove> moves = new ArrayList<>();
		final Map<Suit, List<Card>> foundation = board.getFoundation();
		final List<Column> columns = board.getColumns();
		for (final Suit suit : Suit.values()) {
			final List<Card> foundationForSuit = foundation.get(suit);
			if (!foundationForSuit.isEmpty()) {
				final Card topOfFoundation = foundationForSuit.get(foundationForSuit.size() - 1);
				final List<Card> run = Collections.singletonList(topOfFoundation);
				final ListIterator<Column> iter = columns.listIterator();
				while (iter.hasNext()) {
					final int i = iter.nextIndex();
					final Column column = iter.next();
					if (column.canAddRun(run)) {
						moves.add(new FoundationToColumnMove(i, topOfFoundation));
					}
				}
			}
		}
		return moves;
	}
    
    /**
     * The index in the board of the destination column for the card.
     */
    private final int destinationColumnIndex;
    
    /**
     * The card being moved.
     */
    private final Card card;
    
    /**
     * Creates a move that takes a card from the Foundation and puts it on top
     * of a column.
     * 
     * @param destinationColumnIndex the index of the column into the board
     * @param card the card being moved
     */
    public FoundationToColumnMove(final int destinationColumnIndex, final Card card) {
        super();
        this.destinationColumnIndex = destinationColumnIndex;
        this.card = card;
    }
    
    /**
     * Creates a move that takes a card from the Foundation and puts it on top
     * of a column.
     * 
     * @param destinationColumnIndex the index of the column into the board
     * @param suit the suit of card to pull from the foundation
     * @param board the board to get the card from
     * @throws IndexOutOfBoundsException if the board foundation has no cards for the suit
     */
    public FoundationToColumnMove(final int destinationColumnIndex, final Suit suit, final Board board) {
    	this(destinationColumnIndex, board.getTopOfFoundation(suit));
    }
    
    /**
     * Returns the card being moved.
     * 
     * @return the card moved
     * @see #getCards()
     */
    public Card getCard() {
        return card;
    }
    
    @Override
    public boolean hasCards() {
        return true;
    }
    
    @Override
    public List<Card> getCards() {
        return Collections.singletonList(card);
    }

    @Override
    public boolean isStockPileModification() {
        return false;
    }

    @Override
    public boolean isFromStockPile() {
        return false;
    }

    @Override
    public boolean isFromFoundation() {
        return true;
    }

    @Override
    public boolean isFromColumn() {
        return false;
    }

    @Override
    public boolean isFromColumn(int columnIndex) {
        return false;
    }

    @Override
    public int getSourceColumnIndex() {
        throw new IllegalStateException("Not a move from a column.");
    }

    @Override
    public boolean isToFoundation() {
        return false;
    }

    @Override
    public boolean isToColumn() {
        return true;
    }

    @Override
    public boolean isToColumn(int columnIndex) {
        return columnIndex == destinationColumnIndex;
    }

    @Override
    public int getDestinationColumnIndex() {
        return destinationColumnIndex;
    }

	@Override
    public Board apply(final Board board) {
        final Pair<Card, Map<Suit, List<Card>>> pair = board.extractCardFromFoundation(card.getSuit());
		final Card card = pair.getFirst();
		final Map<Suit, List<Card>> newFoundation = pair.getSecond();
		
		final List<Column> columns = board.getColumns();
		final Column column = columns.get(destinationColumnIndex);
		final Column newColumn = column.withCard(card);
		
		final List<Column> newColumns = new ArrayList<>(columns);
		newColumns.set(destinationColumnIndex, newColumn);
		
		return new Board(newColumns, board.getStockPile(), board.getStockPileIndex(), newFoundation);
    }
    
    @Override
    public int hashCode() {
        return Integer.rotateLeft(destinationColumnIndex, 8) ^ card.hashCode();
    }
    
    @Override
    public boolean equals(final Object obj) {
        if (obj instanceof FoundationToColumnMove) {
            final FoundationToColumnMove move = (FoundationToColumnMove) obj;
            return destinationColumnIndex == move.destinationColumnIndex && card.equals(move.card);
        } else {
            return false;
        }
    }
    
    @Override
    public String toString() {
        return "Move " + card + " from foundation to column " + destinationColumnIndex + ".";
    }
    
}

File: csharp/Solitaire/Klondike/Move/FoundationToColumnMove.cs
﻿using Silnith.Game.Deck;
using System;
using System.Collections.Generic;

namespace Silnith.Game.Klondike.Move
{
    /// <summary>
    /// A move that takes a card from the foundation and puts it onto a column run.
    /// </summary>
    /// <remarks>
    /// <para>
    /// If the column is empty, only a <see cref="Value.King"/> may be
    /// moved.  If the column is not empty, then the card must adhere to the rules of
    /// a run.
    /// </para>
    /// </remarks>
    public class FoundationToColumnMove : ISolitaireMove, IEquatable<FoundationToColumnMove?>
    {
        /// <summary>
        /// The equivalent of the Java values() provided for enumerations.
        /// </summary>
        private static readonly IEnumerable<Suit> suits = new List<Suit>()
        {
            Suit.Club,
            Suit.Diamond,
            Suit.Heart,
            Suit.Spade,
        };

        /// <summary>
        /// Finds all moves of a card from the foundation to a column run.
        /// </summary>
        /// <remarks>
        /// <para>
        /// </para>
        /// </remarks>
        /// <param name="board">The board to examine.</param>
        /// <returns>An enumerable of moves.</returns>
        public static IEnumerable<FoundationToColumnMove> FindMoves(Board board)
        {
            List<FoundationToColumnMove> moves = new List<FoundationToColumnMove>();
            foreach (Suit suit in suits)
            {
                IReadOnlyList<Card> foundationForSuit = board.Foundation[suit];
                int suitCount = foundationForSuit.Count;
                if (suitCount > 0)
                {
                    Card topOfFoundation = foundationForSuit[suitCount - 1];
                    List<Card> run = new List<Card>()
                    {
                        topOfFoundation,
                    };
                    for (int i = 0; i < board.Columns.Count; i++)
                    {
                        Column column = board.Columns[i];
                        if (column.CanAddRun(run))
                        {
                            moves.Add(new FoundationToColumnMove(i, topOfFoundation));
                        }
                    }
                }
            }
            return moves;
        }

        /// <summary>
        /// The card being moved.
        /// </summary>
        public Card Card
        {
            get;
        }

        /// <inheritdoc/>
        public IReadOnlyList<Card> Cards
        {
            get;
        }

        /// <inheritdoc/>
        public int SourceColumnIndex => throw new ArgumentException("Not a move from a column.");

        /// <inheritdoc/>
        public int DestinationColumnIndex
        {
            get;
        }

        /// <inheritdoc/>
        public bool HasCards => true;

        /// <inheritdoc/>
        public bool IsStockPileModification => false;

        /// <inheritdoc/>
        public bool IsStockPileAdvance => false;

        /// <inheritdoc/>
        public bool IsStockPileRecycle => false;

        /// <inheritdoc/>
        public bool IsFromStockPile => false;

        /// <inheritdoc/>
        public bool IsFromFoundation => true;

        /// <inheritdoc/>
        public bool IsFromColumn => false;

        /// <inheritdoc/>
        public bool IsToFoundation => false;

        /// <inheritdoc/>
        public bool IsToColumn => true;

        /// <summary>
        /// Constructs a move that takes a card from the foundation and puts it on top
        /// of a column.
        /// </summary>
        /// <param name="destinationColumnIndex">The index of the column into the board.</param>
        /// <param name="card">The card being moved.</param>
        public FoundationToColumnMove(int destinationColumnIndex, Card card)
        {
            DestinationColumnIndex = destinationColumnIndex;
            Card = card;
            Cards = new List<Card>()
            {
                card,
            };
        }

        /// <summary>
        /// Constructs a move that takes a card from the foundation and puts it on top
        /// of a column.
        /// </summary>
        /// <param name="destinationColumnIndex">The index of the column into the board.</param>
        /// <param name="suit">The suit of card to pull from the foundation.</param>
        /// <param name="board">The board to get the card from.</param>
        /// <exception cref="ArgumentOutOfRangeException">If the board foundation has no cards for <paramref name="suit"/>.</exception>
        public FoundationToColumnMove(int destinationColumnIndex, Suit suit, Board board) : this(destinationColumnIndex, board.GetTopOfFoundation(suit))
        {
        }

        /// <inheritdoc/>
        public bool TakesFromColumn(int columnIndex) => false;

        /// <inheritdoc/>
        public bool AddsToColumn(int columnIndex) => columnIndex == DestinationColumnIndex;

        /// <inheritdoc/>
        public Board Apply(Board board)
        {
            Tuple<Card, IReadOnlyDictionary<Suit, IReadOnlyList<Card>>> tuple = board.ExtractCardFromFoundation(Card.Suit);
            Card card = tuple.Item1;
            IReadOnlyDictionary<Suit, IReadOnlyList<Card>> newFoundation = tuple.Item2;

            IReadOnlyList<Column> newColumns = new List<Column>(board.Columns)
            {
                [DestinationColumnIndex] = board.Columns[DestinationColumnIndex].WithCard(card),
            };

            return new Board(newColumns, board.StockPile, board.StockPileIndex, newFoundation);
        }

        /// <inheritdoc/>
        public override bool Equals(object? obj)
        {
            return Equals(obj as FoundationToColumnMove);
        }

        /// <inheritdoc/>
        public bool Equals(FoundationToColumnMove? other)
        {
            return other != null
                && DestinationColumnIndex == other.DestinationColumnIndex
                && Card.Equals(other.Card);
        }

        /// <inheritdoc/>
        public override int GetHashCode()
        {
            return HashCode.Combine(DestinationColumnIndex, Card);
        }

        /// <inheritdoc/>
        public override string ToString()
        {
            return $"Move {Card} from foundation to column {DestinationColumnIndex}.";
        }
    }
}
